In [105]:
import pandas as pd
import numpy as np

### Ejercicio III

#### Pregunta 1

In [175]:
#carga datos
d2019=pd.read_csv("d2019.csv",index_col=0)
d2020=pd.read_csv("d2020.txt",index_col=0,sep=" ")

In [176]:
df_sii_con=pd.concat((d2019,d2020),axis=0)
df_sii_con.head()

Día     Ene     Feb     Mar     Abr     May     Jun     Jul     Ago  \
0    1     NaN  657.81  651.79  678.53     NaN     NaN  679.15  700.82   
1    2  694.77     NaN     NaN  672.56  678.71     NaN  677.62  704.88   
2    3  697.09     NaN     NaN  670.05  683.31  709.80  681.30     NaN   
3    4  697.64  655.39  656.57  667.55     NaN  704.59  680.48     NaN   
4    5     NaN  653.80  659.07  666.32     NaN  696.78  677.64  711.13   

      Sep     Oct     Nov     Dic  year  
0     NaN  728.21     NaN     NaN  2019  
1  720.20  729.38     NaN  812.13  2019  
2  724.72  727.80     NaN  804.29  2019  
3  725.69  719.53  735.05  802.44  2019  
4  723.30     NaN  736.66  796.75  2019

In [177]:
#melt o unpivot de la tabla
df_sii=pd.melt(df_sii_con,id_vars=['Día','year'],var_name ='mes').iloc[:,[1,2,0,3]]
df_sii.columns=["year","mes","dia","valor"]

In [178]:
df_sii.head(5)

year  mes  dia   valor
0  2019  Ene    1     NaN
1  2019  Ene    2  694.77
2  2019  Ene    3  697.09
3  2019  Ene    4  697.64
4  2019  Ene    5     NaN

In [179]:
#guardar como csv
df_sii.to_csv("df_sii.csv")

#### Pregunta 2

In [180]:
df_sii.groupby("mes",as_index = True)["valor"].agg("mean").head()

mes
Abr    760.389048
Ago    749.182619
Dic    752.561750
Ene    724.854773
Feb    726.342000
Name: valor, dtype: float64

In [181]:
df_sii.groupby(["year","mes"],as_index = False)["valor"].agg("mean").head()

year  mes       valor
0  2019  Abr  667.399048
1  2019  Ago  713.703333
2  2019  Dic  770.390500
3  2019  Ene  677.061818
4  2019  Feb  656.304500

##### 
Lo que cambia usando as_index=False es que el resultado es un data frame, con True es una serie. <br>
Cuando se agrupa por mes calcula la media de los meses en los dos años, mientras que agrupando por año y mes sólo calcula la media del mes en un año

#### Pregunta 3

##### 
Respecto a los valores nulos son días feriados y fines de semana. En dichos días no se cotiza el dolar.

In [182]:
#Eliminación de fechas inexistentes en calendario
#Febrero
df_sii=df_sii.drop(df_sii[(df_sii["mes"]=="Feb")&(df_sii["dia"]>=29)&(df_sii["year"]==2019)].index)
df_sii=df_sii.drop(df_sii[(df_sii["mes"]=="Feb")&(df_sii["dia"]>=30)&(df_sii["year"]==2020)].index) #año bisiesto
#Abril
df_sii=df_sii.drop(df_sii[(df_sii["mes"]=="Abr")&(df_sii["dia"]>=31)&(df_sii["year"]==2019)].index)
df_sii=df_sii.drop(df_sii[(df_sii["mes"]=="Abr")&(df_sii["dia"]>=31)&(df_sii["year"]==2020)].index)
#Junio
df_sii=df_sii.drop(df_sii[(df_sii["mes"]=="Jun")&(df_sii["dia"]>=31)&(df_sii["year"]==2019)].index)
df_sii=df_sii.drop(df_sii[(df_sii["mes"]=="Jun")&(df_sii["dia"]>=31)&(df_sii["year"]==2020)].index)
#Septiembre
df_sii=df_sii.drop(df_sii[(df_sii["mes"]=="Sep")&(df_sii["dia"]>=31)&(df_sii["year"]==2019)].index)
df_sii=df_sii.drop(df_sii[(df_sii["mes"]=="Sep")&(df_sii["dia"]>=31)&(df_sii["year"]==2020)].index)
#Noviembre
df_sii=df_sii.drop(df_sii[(df_sii["mes"]=="Nov")&(df_sii["dia"]>=31)&(df_sii["year"]==2019)].index)
df_sii=df_sii.drop(df_sii[(df_sii["mes"]=="Nov")&(df_sii["dia"]>=31)&(df_sii["year"]==2020)].index)

In [183]:
#creacion campos month y day

df_sii["month"] = np.where(df_sii['mes']=="Ene", 1,
                   np.where(df_sii['mes']=="Feb", 2,
                   np.where(df_sii['mes']=="Mar", 3,
                   np.where(df_sii['mes']=="Abr", 4,
                   np.where(df_sii['mes']=="May", 5,
                    np.where(df_sii['mes']=="Jun", 6,
                    np.where(df_sii['mes']=="Jul", 7,
                    np.where(df_sii['mes']=="Ago", 8,
                    np.where(df_sii['mes']=="Sep", 9,
                    np.where(df_sii['mes']=="Oct", 10,
                    np.where(df_sii['mes']=="Nov", 11,12)))))))))))    

df_sii["day"]=df_sii["dia"]

In [184]:
#creacion campo fechas
df_sii["fecha"]=pd.to_datetime(df_sii[['year', 'month', 'day']])

In [185]:
#creacion campo dia_de_la_semana
df_sii["dia_de_la_semana"]=df_sii['fecha'].dt.day_name()
df_sii["dia_de_la_semana"]=np.where(df_sii["dia_de_la_semana"]=="Monday","Lunes",
                           np.where(df_sii["dia_de_la_semana"]=="Tuesday","Martes",
                           np.where(df_sii["dia_de_la_semana"]=="Wednesday","Miercoles",
                           np.where(df_sii["dia_de_la_semana"]=="Thursday","Jueves",
                           np.where(df_sii["dia_de_la_semana"]=="Friday","Viernes",
                           np.where(df_sii["dia_de_la_semana"]=="Saturday","Sabado","Domingo"))))))

df_sii=df_sii[["year","mes","dia","valor","dia_de_la_semana"]]
df_sii

year  mes  dia   valor dia_de_la_semana
0    2019  Ene    1     NaN           Martes
1    2019  Ene    2  694.77        Miercoles
2    2019  Ene    3  697.09           Jueves
3    2019  Ene    4  697.64          Viernes
4    2019  Ene    5     NaN           Sabado
..    ...  ...  ...     ...              ...
739  2020  Dic   27     NaN          Domingo
740  2020  Dic   28  710.26            Lunes
741  2020  Dic   29  710.64           Martes
742  2020  Dic   30  711.24        Miercoles
743  2020  Dic   31     NaN           Jueves

[731 rows x 5 columns]

#### Pregunta 4

In [273]:
#creación de un df para cada hoja del archivo excel
dolar_observado_bc = pd.read_excel("https://raw.githubusercontent.com/hfas85/tarea1-add/master/data/raw/dolar_observado_bc.xlsx",
                                 skiprows=2, dtype={'Periodo':str }, sheet_name=None)

df_1980 = dolar_observado_bc.get('80')
df_1990 = dolar_observado_bc.get('90')
df_2000 = dolar_observado_bc.get('00')
df_2010 = dolar_observado_bc.get('10')
df_2020 = dolar_observado_bc.get('20')

df_1980["Periodo"]=df_1980["Periodo"].str[0:10]
df_1990["Periodo"]=df_1990["Periodo"].str[0:10]
df_2000["Periodo"]=df_2000["Periodo"].str[0:10]
df_2010["Periodo"]=df_2010["Periodo"].str[0:10]
df_2020["Periodo"]=df_2020["Periodo"].str[0:10]


In [274]:
#creación de campos year,mes,dia para cada df
df_1980[["year", "mes", "dia"]] = df_1980["Periodo"].str.split("-", expand = True)
df_1990[["year", "mes", "dia"]] = df_1990["Periodo"].str.split("-", expand = True)
df_2000[["year", "mes", "dia"]] = df_2000["Periodo"].str.split("-", expand = True)
df_2010[["year", "mes", "dia"]] = df_2010["Periodo"].str.split("-", expand = True)
df_2020[["year", "mes", "dia"]] = df_2020["Periodo"].str.split("-", expand = True)

df_1980.columns = ["Periodo","Dolar","year","mes","dia"]
df_1990.columns = ["Periodo","Dolar","year","mes","dia"]
df_2000.columns = ["Periodo","Dolar","year","mes","dia"]
df_2010.columns = ["Periodo","Dolar","year","mes","dia"]
df_2020.columns = ["Periodo","Dolar","year","mes","dia"]

df_2020.head()

Periodo   Dolar  year mes dia
0  2020-01-01     NaN  2020  01  01
1  2020-01-02  748.74  2020  01  02
2  2020-01-03  754.16  2020  01  03
3  2020-01-04     NaN  2020  01  04
4  2020-01-05     NaN  2020  01  05

In [276]:
#creacion de un solo data frame
df_bc_all=pd.concat((df_1980,df_1990,df_2000,df_2010,df_2020),axis=0)
df_bc_all

Periodo   Dolar  year mes dia
0    1982-08-09   55.65  1982  08  09
1    1982-08-10   62.79  1982  08  10
2    1982-08-11   64.07  1982  08  11
3    1982-08-12   60.52  1982  08  12
4    1982-08-13   56.87  1982  08  13
..          ...     ...   ...  ..  ..
969  2022-08-27     NaN  2022  08  27
970  2022-08-28     NaN  2022  08  28
971  2022-08-29  900.22  2022  08  29
972  2022-08-30  890.90  2022  08  30
973  2022-08-31  882.11  2022  08  31

[14633 rows x 5 columns]

In [286]:
# Creación de csv agrupando por "year"
for (year), group in df_bc_all.groupby(['year']):
   group.to_csv(f'{year}.csv', index=False)


In [289]:
#importar csv de los años 2019 y 2020
d_2019_2=pd.read_csv("2019.csv")
d_2020_2=pd.read_csv("2020.csv")

df_bc_2019_2020=pd.concat((d_2019_2,d_2020_2),axis=0)

Periodo   Dolar  year  mes  dia
0  2019-01-01     NaN  2019    1    1
1  2019-01-02  694.77  2019    1    2
2  2019-01-03  697.09  2019    1    3
3  2019-01-04  697.64  2019    1    4
4  2019-01-05     NaN  2019    1    5

In [296]:
#comparación df_bc_2019_2020 con df_sii.
df_bc_2019_2020.head(10)

Periodo   Dolar  year  mes  dia
0  2019-01-01     NaN  2019    1    1
1  2019-01-02  694.77  2019    1    2
2  2019-01-03  697.09  2019    1    3
3  2019-01-04  697.64  2019    1    4
4  2019-01-05     NaN  2019    1    5
5  2019-01-06     NaN  2019    1    6
6  2019-01-07  688.42  2019    1    7
7  2019-01-08  679.69  2019    1    8
8  2019-01-09  680.10  2019    1    9
9  2019-01-10  678.09  2019    1   10

In [297]:
#comparación df_bc_2019_2020 con df_sii.
df_sii.head(10)

year  mes  dia   valor dia_de_la_semana
0  2019  Ene    1     NaN           Martes
1  2019  Ene    2  694.77        Miercoles
2  2019  Ene    3  697.09           Jueves
3  2019  Ene    4  697.64          Viernes
4  2019  Ene    5     NaN           Sabado
5  2019  Ene    6     NaN          Domingo
6  2019  Ene    7  688.42            Lunes
7  2019  Ene    8  679.69           Martes
8  2019  Ene    9  680.10        Miercoles
9  2019  Ene   10  678.09           Jueves

In [298]:
#comparación df_bc_2019_2020 con df_sii.
df_bc_2019_2020.shape, df_sii.shape

((731, 5), (731, 5))

##### 
Los valores de los DF son iguales, la cantidad de registros en ambos es de 731

#### Pregunta 5

#### Pregunta 6